apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableServiceApiClientPartial.cs

```diff
@@ -133,4 +187,383 @@ public partial class BigtableServiceApiClientImpl
+            public static readonly ChannelOptionComparer Instance = new ChannelOptionComparer();
+
+            private ChannelOptionComparer() { }
+
+            public bool Equals(ChannelOption x, ChannelOption y)
+            {
+                if (x == null && y == null)
+                {
+                    return true;
+                }
+
+                if (x == null || y == null)
+                {
+                    return false;
+                }
+
+                if (x.Type != y.Type || x.Name != y.Name)
+                {
+                    return false;
+                }
+
+                switch (x.Type)
+                {
+                    case ChannelOption.OptionType.Integer:
+                        return x.IntValue == y.IntValue;
+
+                    case ChannelOption.OptionType.String:
+                        return y.StringValue == y.StringValue;
+
+                    default:
+                        throw new ArgumentException("Unexpected channel option type: " + x.Type);
+                }
+            }
+
+            public int GetHashCode(ChannelOption obj)
+            {
+                return
+                    obj == null ? 0 :
+                    obj.Type == ChannelOption.OptionType.Integer ?
+                        EqualityHelpers.CombineHashCodes(obj.IntValue, obj.Name.GetHashCode(), (int)obj.Type) :
+                    obj.Type == ChannelOption.OptionType.String ?
+                        EqualityHelpers.CombineHashCodes(obj.StringValue.GetHashCode(), obj.Name.GetHashCode(), (int)obj.Type) :
+                    throw new ArgumentException("Unexpected channel option type: " + obj.Type);
+            }
+        }
+
+        private struct GcpCallInvokerKey : IEquatable<GcpCallInvokerKey>
+        {
+            public readonly ServiceEndpoint Endpoint;
+            public readonly List<ChannelOption> Options;
+            public readonly int _hash;
+
+            public GcpCallInvokerKey(ServiceEndpoint endpoint, List<ChannelOption> options)
+            {
+                Endpoint = endpoint;
+                Options = options;
+                _hash = EqualityHelpers.CombineHashCodes(
+                    endpoint.GetHashCode(),
+                    EqualityHelpers.GetListHashCode(options, ChannelOptionComparer.Instance));
+            }
+
+            public override int GetHashCode() => _hash;
+
+            public override bool Equals(object obj) => obj is GcpCallInvokerKey other && Equals(other);
+
+            public bool Equals(GcpCallInvokerKey other) =>
+                Endpoint.Equals(other.Endpoint) && Options.SequenceEqual(other.Options, ChannelOptionComparer.Instance);
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="endpoint"></param>
+        /// <param name="options"></param>
+        /// <returns></returns>
+        public GcpCallInvoker GetGcpCallInvoker(ServiceEndpoint endpoint, IEnumerable<ChannelOption> options)
```

Comment:
So presumably if we wanted to be able to generate code for an API not on GCP, it could call the existing methods, right?

Reply:
Yes, that's the idea. I guess if this makes the ChannelPool too closely tied to GCP and that worries you, we can even put this stuff in a totally new class. The only downside there is it would mean more changes would be necessary for the auto-generation, since the s_channelPool member would need to have a different type.

Reply:
It sounds like we may need code-gen to be aware of whether an API is a Cloud API or not. I think that's reasonable, to be honest.

cc @chrisdunelm as he'll be the one doing the work :)

Reply:
Ok sounds good